### DISCUSSION TITLE
How do I resize an image and preserving the aspect ratio lossless?

### DISCUSSION THREAD

**Discussion by Bluebuster116** [ImageMagick/ImageMagick#5109] [Consulting]
Hi. I am still a novice with Imagemagick. I've been using it for a while but I can't recall how to resize the image preserving the aspect ratio.

Let's say how do I make an image so that it has a width of 200px?

Conversely how to do the same but with a height of 200px?

Will those functions use Lanczos algorithm or something? Is this something that I need to include in the code? I ask this to reassure that the result is lossless.

To know my version I used this:

`convert -version`

The result was that:
I am using this version: ImageMagick 6.9.3-2 Q16 x86 2016-01-23

Because of this, can you please reply using a syntax which my version for Windows 7 can understand?

Can you please help me here?. Thanks in advance.

**Comment by fmw42** [comment]
Width 200:  `convert image.suffix -resize 200x output.suffix`
Height 200: `convert image.suffix -resize x200 output.suffix`

Default for downsizing is Lanczos.

See 

https://legacy.imagemagick.org/Usage/resize/
https://legacy.imagemagick.org/Usage/filter/
https://imagemagick.org/script/command-line-processing.php#geometry

**Reply by fmw42** [reply]
For novices, see

http://www.imagemagick.org/discourse-server/viewtopic.php?f=1&t=9620
http://www.imagemagick.org/script/command-line-processing.php
https://imagemagick.org/script/command-line-options.php
http://www.imagemagick.org/Usage/reference.html
http://www.imagemagick.org/Usage/
https://github.com/ImageMagick/usage-markdownhttps://imagemagick.org/script/porting.php#cli
https://legacy.imagemagick.org/discourse-server/index.php

**Comment by snibgo** [comment]
> Will those functions use Lanczos algorithm or something? Is this something that I need to include in the code? I ask this to reassure that the result is lossless.

What do you mean by "lossless"? "-resize" will calculate each output pixel from some number of input pixels. "-sample" will set each output pixel from only one input pixel.

**Comment by fmw42** [comment]
Also do not save as JPG. JPG is a lossy compression format. Use PNG or TIFF

**Comment by Bluebuster116** [comment]
@fmw42

Thanks for that example. Just to confirm, the blank space after the x tells imagemagick to automatically scale the image only the specified edge of the image, right?

I've read the links you mentioned. However what if I want to select another algorithm?. How can I choose that? Can you put an example for this please?

By the way in your comment you also suggested me not to use jpg but instead png. Yes that's what I've been doing but, by how much does the image get distortion when resized by using jpg?

@snibgo

>What do you mean by "lossless"? "

Well I was intending to say that the output image doesn't look too grainy, if you know my mean.

By reading at the documentation I found the part where it says Adaptative resize produces images a little bit better. But it also mentions that produces aliasing. What is it?. This I don't understand well.

In addition to this I'd like to know. Is there a way to turn a word document or excel sheet page into a png or jpg using my imagemagick version? 

Also, can it be resized? Let's say that the page I want to convert is in a landscape orientation, can it be resized to fit the width to 640px?. What if it is in a vertical orientation and fit to 320px? Is it possible to do that?. 

I mean in a single command for both separately, to read the document which I know it is in landscape and that to resize it as 640px for example.

Can you please tell me an example with the syntax?. It would be nice to see a screenshot of how it would look like. Can you guys attend these doubts?.

**Comment by fmw42** [comment]
> Thanks for that example. Just to confirm, the blank space after the x tells imagemagick to automatically scale the image only the specified edge of the image, right?

yes, 100x would scale the width to 100 and keep the height in proportion. Likewise x100 would scale the height to 100 and keep the width in proportion.  100x100 would scale by whichever of width or height was the larger dimension in your image to 100 and the other dimensions would be in proportion.

`I've read the links you mentioned. However what if I want to select another algorithm?. How can I choose that? Can you put an example for this please?`

Sorry, but I do not know what algorithm your want? I do not understand the question!

> By the way in your comment you also suggested me not to use jpg but instead png. Yes that's what I've been doing but, by how much does the image get distortion when resized by using jpg?

**Comment by snibgo** [comment]
> In addition to this I'd like to know. Is there a way to turn a word document or excel sheet page into a png or jpg using my imagemagick version?

IM cannot rasterize Word documents or Excel spreadsheets.

**Comment by Bluebuster116** [comment]
@fmw42

>Sorry, but I do not know what algorithm your want? I do not understand the question!

I don't know what resizing algorithms can Imagemagick handle other than Lanczos. I know there is bicubic. How would your example be if I choose that algorithm? Can you show an example of the syntax for my version?

There are others which come to mind like bilinear, nearest neighbor, fourier, edge directed. However in terms of the one which would be the best all purpose one, this is what I don't know. Will the syntax for nearest neighbor be the same as for the others?

I did found this article about [resampling filters](https://legacy.imagemagick.org/Usage/filter/) I don't understand it very well, but it appears to say that Gaussian filter does produce more smooth resizing when an image is augmented?. Let's say to produce 100px width size thumbnails from larger images, will it be a good idea?  Can you confirm this?. In the example there it is given as a percentage. Does this means that it cannot be used with fixed pixels?.

It appears that imagemagick cannot read word documents or excel sheets. Is this because of my version or is it the current status? Anyways, is there some workaround to solve this problem?

I mean I could create a pdf document from libreoffice or any office application and that I think it could be "read" by imagemagick.

With this in mind, can imagemagick search in that document the page I want and make into a png file?. Then how to do that image with and without a specified width and height?.

Can this also be done with a pdf file?. Let's say can imagemagick "read" a certain page and make that page into a png?

I appreciate a lot if you could attend these doubts.

**Reply by fmw42** [reply]
Add -filter X to your command for resizing.

`convert image.suffix -filter X -resize 200x output.suffix`

where X can be any of the filters on that link you showed or use

```
convert -list filter

Bartlett
Blackman
Bohman
Box
Catrom
Cosine
Cubic
Gaussian
Hamming
Hann
Hermite
Jinc
Kaiser
Lagrange
Lanczos
Lanczos2
Lanczos2Sharp
LanczosRadius
LanczosSharp
Mitchell
Parzen
Point
Quadratic
Robidoux
RobidouxSharp
Sinc
SincFast
Spline
Triangle
Welch

```

to see all the filters available on your version of Imagemagick.

For making thumbnails, I would suggest you use -thumbnail rather than -resize. The argument are the same, but it chooses an optimal filter for reducing and removes the meta data to keep the file size smaller.

If you have Ghostscript on your system on as a delegate to Imagemagick, then Imagemagick will use it to read and rasterize a PDF. So you could convert your Excel or Word document to PDF and then Imagemagick could read it.

`convert -density XX image.pdf image.png`

where XX is the density you want to read your PDF. The larger the density the large the rasterized PNG dimensions will be and the higher the quality. XX=72 is the default.

You can then do


`convert -density 288 image.pdf -resize 25% image.png`

or

`convert -density 288 image.pdf -resize WxH image.png`

where WxH is the size you want for the output. For example WxH = 595x842 is the size you would want for A4 paper size.

As far as I know, Imagemagick cannot currently read Word or Excel files unless you either export as PDF or simple text or do a screen snap to make it into a raster image such as PNG or TIFF.
